src/pretty/flow_like.rs

```diff
@@ -0,0 +1,114 @@
+use pretty::{Arena, DocAllocator};
+use typst_syntax::{SyntaxKind, SyntaxNode};
+
+use super::{util::is_comment_node, ArenaDoc, PrettyPrinter};
+
+pub struct FlowItem<'a>(Option<FlowItemRepr<'a>>);
+
+struct FlowItemRepr<'a> {
+    doc: ArenaDoc<'a>,
+    /// Whether a space can be added before the doc. Not enforcement.
+    space_before: bool,
+    ///  Whether a space can be added after the doc.
+    space_after: bool,
+}
+
+impl<'a> FlowItem<'a> {
+    pub fn doc(doc: ArenaDoc<'a>, space_after: bool) -> Self {
+        Self(Some(FlowItemRepr {
+            doc,
+            space_before: true,
+            space_after,
+        }))
+    }
+
+    pub fn doc_spc(doc: ArenaDoc<'a>) -> Self {
+        Self::doc(doc, true)
+    }
+
+    pub fn tight_doc(doc: ArenaDoc<'a>, space_after: bool) -> Self {
+        Self(Some(FlowItemRepr {
+            doc,
+            space_before: false,
+            space_after,
+        }))
+    }
```

Comment:
These constructors are confusing..

Reply:
Consider adding a constructor `new` that takes doc, space_before, space_after. And implement rest of them using `new`

Reply:
pls also document the difference between all of them. 🙇‍

Reply:
BTW, do you think it is a bad idea to just return Option((Doc, bool, bool))?

Reply:
pls dont, it's worse, because we dont know what these bools mean now